nexus-repository-conan/src/main/java/org/sonatype/repository/conan/internal/hosted/v1/ConanHostedFacet.java

```diff
@@ -138,12 +141,28 @@ protected void doPutArchive(final ConanCoords coord,
     saveAsset(tx, asset, tempBlob);
   }
 
+  @TransactionalTouchBlob
   public String getDownloadUrlAsJson(final ConanCoords coords) throws JsonProcessingException {
-    String repositoryUrl = getRepository().getUrl();
     if (StringUtils.isEmpty(coords.getSha())) {
-      return generateDownloadUrlsAsJson(coords, repositoryUrl);
+      return generateDownloadUrlsAsJson(coords);
     }
-    return generateDownloadPackagesUrlsAsJson(coords, repositoryUrl);
+    return generateDownloadPackagesUrlsAsJson(coords);
+  }
+
+  public String getUploadUrlAsJson(final ConanCoords coords, final Set<String> assetsToUpload)
+      throws JsonProcessingException
+  {
+    String repositoryUrl = getRepository().getUrl();
+    Map<String, String> downloadUrls = assetsToUpload.stream()
+        .map(fileName -> Arrays.stream(AssetKind.values())
+            .filter(assetKind -> assetKind.getFilename().equals(fileName))
+            .findAny()
+            .orElseThrow(
+                () -> new IllegalArgumentException("Unknown asset kind for uploading file: " + fileName))
+        )
+        .collect(Collectors.toMap(AssetKind::getFilename,
+            x -> repositoryUrl + "/" + ConanHostedHelper.getHostedAssetPath(coords, x)));
```

Comment:
Minor: I'd find a for loop more readable over nested streams.
```
for (String fileName : assetsToUpload) {
  AssetKind assetKind = AssetKind.valueFromFileName(fileName);
  if (assetKind == null) {
    throw new IllegalArgumentException("Unknown asset kind for uploading file: " + fileName);
  }
  downloadUrls.put(fileName, repositoryUrl + "/" + ConanHostedHelper.getHostedAssetPath(coords, assetKind));
}
```
Adding a method to `AssetKind` to do the conversation between filename and `AssetKind`.
```
static AssetKind valueFromFileName(final String filename) {
  values().find { it.filename == filename }
}
```

Reply:
Ok, let use such approach 9007e590aa0043bb8f7f78080657c842661f3c94